node/src/block_producer/block_producer_reducer.rs

```diff
@@ -12,68 +12,104 @@ use mina_p2p_messages::{
+        BlockProducerEnabled::reducer(state_context, action);
     }
 }
 
 impl BlockProducerEnabled {
-    pub fn reducer(
-        &mut self,
-        action: BlockProducerActionWithMetaRef<'_>,
-        best_chain: &[AppliedBlock],
-    ) {
+    pub fn reducer(mut state_context: Substate<State>, action: BlockProducerActionWithMetaRef<'_>) {
         let (action, meta) = action.split();
+        let Ok(global_state) = state_context.get_substate_mut() else {
+            return;
+        };
+
+        let best_chain = &global_state.transition_frontier.best_chain;
+        let Some(state) = global_state.block_producer.as_mut() else {
+            return;
+        };
+
         match action {
             BlockProducerAction::VrfEvaluator(action) => {
-                self.vrf_evaluator.reducer(meta.with_action(action))
+                BlockProducerVrfEvaluatorState::reducer(
+                    Substate::from_compatible_substate(state_context),
+                    meta.with_action(action),
+                );
             }
             BlockProducerAction::BestTipUpdate { best_tip } => {
-                self.injected_blocks.remove(best_tip.hash());
+                state.injected_blocks.remove(best_tip.hash());
                 // set the genesis timestamp on the first best tip update
                 // TODO: move/remove once we can generate the genesis block
-                if self.vrf_evaluator.genesis_timestamp == redux::Timestamp::ZERO {
-                    self.vrf_evaluator.genesis_timestamp = best_tip.genesis_timestamp();
+                if state.vrf_evaluator.genesis_timestamp == redux::Timestamp::ZERO {
+                    state.vrf_evaluator.genesis_timestamp = best_tip.genesis_timestamp();
+                }
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                Self::dispatch_best_tip_update(dispatcher, state, best_tip);
+            }
+            BlockProducerAction::WonSlotSearch => {
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                if let Some(won_slot) = state.block_producer.with(None, |bp| {
+                    let best_tip = state.transition_frontier.best_tip()?;
+                    let cur_global_slot = state.cur_global_slot()?;
+                    bp.vrf_evaluator.next_won_slot(cur_global_slot, best_tip)
+                }) {
+                    dispatcher.push(BlockProducerAction::WonSlot { won_slot });
                 }
             }
-            BlockProducerAction::WonSlotSearch => {}
             BlockProducerAction::WonSlot { won_slot } => {
-                self.current = BlockProducerCurrentState::WonSlot {
+                state.current = BlockProducerCurrentState::WonSlot {
                     time: meta.time(),
                     won_slot: won_slot.clone(),
                 };
+
+                state_context
```

Comment:
same as in the other case, lets not introduce this new pattern and assign `dispatcher` instead.